### PR TITLE
make sign-up function

### PR DESCRIPTION
**Pull request by VuongLong**

### CODE REVIEW

File: app/assets/stylesheets/users.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (phamkykhoi) [inline app/assets/stylesheets/users.scss]: Remove this file

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+class UsersController < ApplicationController
+  
+  def show
+    @user = User.find(params[:id])
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  end
+  
+  def new
+    @user = User.new
+  end
+
+   def create
+    @user = User.new(user_params)
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: I18n

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+      redirect_to @user
+    else
+      render 'new'
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: ' ' => " "

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    end
+  end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:name, :email, :password,
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+      flash[:success] = "Welcome to the Sample App!"
+      redirect_to @user
+    else
+      render 'new'
+    end
+  end
+
+  private
+
+    def user_params
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Indent

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
+      redirect_to @user
+    else
+      render 'new'
+    end
+  end
+
+  private
+
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove line empty

File: app/helpers/users_helper.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+module UsersHelper
+
+  # Returns the Gravatar for the given user.
+  def gravatar_for(user, options = {size: 50})
```
Comment (phamkykhoi) [inline app/helpers/users_helper.rb]: Move 50 to Settings

File: app/models/user.rb [hunk 10]
```diff
@@ -10,4 +10,34 @@ class User < ActiveRecord::Base
...
+  has_secure_password
+  validates :password, presence: true, length: {minimum: Settings.minimum_password_length}, allow_nil: true
+  
+  validate  :picture_size
+
+  private
+
+    # Validates the size of an uploaded picture.
+    def picture_size
+      if picture.size > 5.megabytes
```
Comment (phamkykhoi) [inline app/models/user.rb]: Move 5 to Settings

File: app/models/user.rb [hunk 11]
```diff
@@ -10,4 +10,34 @@ class User < ActiveRecord::Base
...
+        errors.add(:picture, "should be less than 5MB")
```
Comment (phamkykhoi) [inline app/models/user.rb]: I18n

File: app/uploaders/picture_uploader.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
+# encoding: utf-8
+
+class PictureUploader < CarrierWave::Uploader::Base
+
+  # Include RMagick or MiniMagick support:
+  # include CarrierWave::RMagick
+  # include CarrierWave::MiniMagick
+
+  include CarrierWave::MiniMagick
+  process resize_to_limit: [400, 400]
```
Comment (phamkykhoi) [inline app/uploaders/picture_uploader.rb]: Move 400 to Settings

File: app/uploaders/picture_uploader.rb [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+  # Create different versions of your uploaded files:
+  # version :thumb do
+  #   process :resize_to_fit => [50, 50]
+  # end
+
+  # Add a white list of extensions which are allowed to be uploaded.
+  # For images you might use something like this:
+  def extension_white_list
+    %w(jpg jpeg gif png)
```
Comment (phamkykhoi) [inline app/uploaders/picture_uploader.rb]: Move jpg jpeg gif png to Settings

File: app/views/users/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+      <%= f.email_field :email, class: 'form-control' %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: 'form-control' %>
+
+      <span class="picture">
+    	<%= f.file_field :picture %>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: Indent

File: app/views/users/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+  	  </span>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: Indent

File: config/routes.rb [hunk 16]
```diff
@@ -3,4 +3,61 @@
   get "help"    =>  "static_pages#help"
   get "about"   =>  "static_pages#about"
   get "contact" =>  "static_pages#contact"
+
+  get 'signup' => 'users#new'
```
Comment (phamkykhoi) [inline config/routes.rb]: ' ' => " "

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+  def show
+    @user = User.find params[:id] 
+  end
+  
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    if @user.save
+      flash[:success] = t("static_pages.welcome")
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/views/users/new.html.erb [hunk 19]
```diff
@@ -0,0 +1,28 @@
+<% provide(:title, 'Sign up') %>
+<h1>Sign up</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: Remove ( )

File: app/views/users/new.html.erb [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+      <%= f.password_field :password, class: 'form-control' %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: 'form-control' %>
+
+      <span class="picture">
+    	  <%= f.file_field :picture %>
+  	  </span>
+      
+      <%= f.submit "Create my account", class: "btn btn-primary" %>
```
Comment (phamkykhoi) [inline app/views/users/new.html.erb]: I18n

File: app/helpers/users_helper.rb [hunk 21]
```diff
@@ -0,0 +1,16 @@
+module UsersHelper
+
+  # Returns the Gravatar for the given user.
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: remove all remark

File: app/helpers/users_helper.rb [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+  def gravatar_for(user, options = {size: Settings.image_size})
```
Comment (dieunb) [inline app/helpers/users_helper.rb]: remove `()`

File: app/helpers/users_helper.rb [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    size = options[:size]
+    
+    if user.picture?
```
Comment (dieunb) [inline app/helpers/users_helper.rb]:
```
+    gravatar_url = if user.picture? 
+      user.picture.url
+    else 
+      "https://secure.gravatar.com/avatar/#{gravatar_id}" 
+    end
```

File: app/uploaders/picture_uploader.rb [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+  # Include RMagick or MiniMagick support:
+  # include CarrierWave::RMagick
+  # include CarrierWave::MiniMagick
+
+  include CarrierWave::MiniMagick
+  process resize_to_limit: [Settings.image_resize, Settings.image_resize]
+
+  if Rails.env.production?
+      storage :fog
+    else
```
Comment (dieunb) [inline app/uploaders/picture_uploader.rb]: indent

File: app/uploaders/picture_uploader.rb [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+  # Include RMagick or MiniMagick support:
+  # include CarrierWave::RMagick
+  # include CarrierWave::MiniMagick
+
+  include CarrierWave::MiniMagick
+  process resize_to_limit: [Settings.image_resize, Settings.image_resize]
+
+  if Rails.env.production?
+      storage :fog
```
Comment (dieunb) [inline app/uploaders/picture_uploader.rb]: indent

File: app/uploaders/picture_uploader.rb [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    else
+      storage :file
```
Comment (dieunb) [inline app/uploaders/picture_uploader.rb]: indent

File: app/models/user.rb [hunk 27]
```diff
@@ -39,4 +54,15 @@ def authenticated? remember_token
   def forget
     update_attributes remember_digest: nil
   end
+  
+  private
+  def picture_size
+    if picture.size > Settings.image_storage.megabytes
+      errors.add(:picture, t("sign_up.warning"))
```
Comment (dieunb) [inline app/models/user.rb]: remove `()`

**Pull request merged**